README.md
# LOCALSTORES 多端缓存方案支持 - (小程序,h5,pc)
### 改进的离线存储
localStores 是一个 JavaScript 库，通过简单类似 localStorage API 的异步存储来改进你的 Web 应用程序的离线体验。它能存储多种类型的数据，而不仅仅是字符串. 支持 IndexedDb 或 WebSql 以及 localStorage.

#  安装
使用localStores, 请下载最新版本 或使用npm (npm insall localstores) 进行安装.

# 数据API
获取或设置离线仓库中的数据的API

# GETITEM
getItem(key);

从仓库中获取 key 对应的值并将结果提供给回调函数。如果 key 不存在，getItem() 将返回 null. 返回为 Promise 对象.

# SETITEM

setItem(key, value, successCallback)

将数据保存到离线仓库.

# 设置 API 

## SETDRIVER
setDriver(driverName)

若可用，强制设置特定的驱动。

* 默认情况下，LocalStore 按照以下顺序选择数据仓库的后端驱动：
  * 1、IndexedDB
  * 2、WebSQL
  * 3、localStorage

如果你想强制使用特定的驱动，可以使用 setDriver()，参数为以下的某一个或多个
* LocalStore.INDEXEDDB
* LocalStore.WEBSQL
* LocalStore.LOCALSTORAGE

## CONFIG
config(options)

设置 LocalStore 选项。在调用 LocalStore 前必先调用它，但可以在 LocalStore 被加载后调用。使用此方法设置的任何配置值都将保留，即使在驱动更改后，所以你也可以先调用 config() 再次调用 setDriver()。以下配置值可以设置：


driver

要使用的首选驱动。与上面的 setDriver 的值格式相同。
默认值：[localforage.INDEXEDDB, localforage.WEBSQL, localforage.LOCALSTORAGE]

name

数据库的名称。可能会在在数据库的提示中会出现。一般使用你的应用程序的名字。在 localStorage 中，它作为存储在 localStorage 中的所有 key 的前缀。
默认值：'localforage

size

数据库的大小（以字节为单位）。现在只用于WebSQL。 默认值：4980736

storeName

数据仓库的名称。在 IndexedDB 中为 dataStore，在 WebSQL 中为数据库 key/value 键值表的名称。仅含字母和数字和下划线。任何非字母和数字字符都将转换为下划线。
默认值：'keyvaluepairs'

version

数据库的版本。将来可用于升级; 目前未使用。
默认值：1.0

description

数据库的描述，一般是提供给开发者的。
默认值：''


# 对 websql 的扩展方法支持.


对使用者需要掌握一定的 Sql 语法能力.

websql 使用示例:
```
import React,{useEffect} from 'react';
import LocalStore from 'localstores';

function App() {
  useEffect(()=>{
    const init = async () =>{
      LocalStore.setDriver(LocalStore.WEBSQL);
      LocalStore.extSetItem('tb_test', {netName:'222333', targetVolume:'test', adjustVolume:'222', lastTargetVolume:'222' },
        '(netName varchar(50),targetVolume varchar(40),adjustVolume varchar(40),lastTargetVolume varchar(40))', function (err) {
          console.log('err', err);
        }, function (err){
          console.log('err', err);
        });
      const rst = await LocalStore.getItem('tb_test');
      console.log('rst',rst);
      LocalStore.updateItem('tb_test',{targetVolume:'test'},{netName:'网点'});
    }
    init();
  },[]);

  return (
    <></>
  );
}

export default App;
```

extSetItem

```
/**
 * 
 * @param {*} tableName 表名 tablename
 * @param {*} tableField 表字段对象 { name : '2'}
 * @param {*} onSuccess 成功回调
 * @param {*} onError 错误回调
*/
function extSetItem(tableName,tableField,createTableField,onSuccess,onError){
}

LocalStore.extSetItem('tb_test', {netName:'222333', targetVolume:'22', adjustVolume:'222', lastTargetVolume:'222' },
    '(netName varchar(50),targetVolume varchar(40),adjustVolume varchar(40),lastTargetVolume varchar(40))', function (err) {
      console.log('err', err);
    }, function (err){
      console.log('err', err);
});
```


removeItem
```
/**
 * 删除表
 * @param {*} key 值 id
 * @param {*} tableName 表名
 * @param {*} callback  function 函数
 * @returns 
 */
function removeItem(key,tableName,callback){

}

LocalStore.removeItem(1,'test',function(){

});

```


updateItem
```
/**
 * 更新数据表
 * @param {*} tableName 表明.
 * @param {*} dbParams  sql 拼接参数.
 * @param {*} dbWhere   sql 条件.
 * @returns promise 对象.
 */
 function updateItem(tableName, dbParams, dbWhere)


 LocalStore.updateItem('tb_test',{targetVolume:'test33333'},{netName:'222333'});
```

truncateItem

```
/**
 * 清空数据表
 * @param {*} tableName 表名
 * @returns 
 */
truncateItem(tableName){}

 LocalStore.truncateItem('tb_adjusttarget22');
```
### 








CHANGELOG.md
# 1.1.0 (2021-04-09)


### Bug Fixes

* modify readme.md ([f8cfcd6](https://github.com/xiangxiong/localStore/commit/f8cfcd6db3308f8eb80f6fe572b4925af8b465b8))


### Features

* init project ([3985bdd](https://github.com/xiangxiong/localStore/commit/3985bdd4d1485ab1eb2b96faf6f4146ea7902798))
* modify project ([71588a0](https://github.com/xiangxiong/localStore/commit/71588a028bd2d0df100ac7d129a8a34ad73f31a0))
* 修改babel.config.js 文件 ([b96c574](https://github.com/xiangxiong/localStore/commit/b96c5743faf998c29dffb634656fff7658ceaf85))
* 修改包名 ([edd2525](https://github.com/xiangxiong/localStore/commit/edd2525a78b0068bd82a656da8c574b706fbb47f))





lib/index.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, global.localstores = factory());
}(this, (function () { 'use strict';

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
  }

  function createDbTable(t, dbInfo, callback, errorCallback) {
    t.executeSql("CREATE TABLE IF NOT EXISTS ".concat(dbInfo.storeName, " ") + '(key varchar(50), value varchar(50))', [], callback, errorCallback);
  }

  function _initStorage$1(options) {
    var self = this;
    var dbInfo = {
      db: null
    };

    if (options) {
      for (var i in options) {
        dbInfo[i] = typeof options[i] !== 'string' ? options[i].toString() : options[i];
      }
    }

    dbInfo.db = openDatabase(String(dbInfo.name), String(dbInfo.version), String(dbInfo.description), dbInfo.size); // dbInfo.db = openDatabase("MySql", "1.0", "我的数据库描述", 1024 * 1024);

    self._dbInfo = dbInfo.db;
    dbInfo.db.transaction(function (t) {
      createDbTable(t, dbInfo, function () {}, function (t, error) {
        console.log(error);
      });
    });
  }
  /**
   * 创建扩展的表
   * @param {*} tableName  扩展表名称
   * @param {*} createTableField  扩展表字段
   * @param {*} onSuccess  成功回调
   * @param {*} onError  错误回调
   * @param {*} context 上下文
   * @returns 
   */


  function createExtTable(tableName, createTableField, onSuccess, onError, context) {
    var extTablePromise = new Promise(function (resolve, reject) {
      try {
        context.transaction(function (ts, result) {
          ts.executeSql("CREATE TABLE IF NOT EXISTS ".concat(tableName, " ") + createTableField, [], onSuccess, onError);
        });
        resolve('success');
      } catch (err) {
        reject(err);
      }
    });
    return extTablePromise;
  }
  /**
   * 
   * @param {*} tableName 表名 tablename
   * @param {*} tableField 表字段对象 { name : '2'}
   * @param {*} onSuccess 成功回调
   * @param {*} onError 错误回调
  */


  function extSetItem(tableName, tableField, createTableField, onSuccess, onError) {
    var _this = this;

    createExtTable(tableName, createTableField, onSuccess, onError, this._dbInfo).then(function (item) {
      if (!tableField) throw new Error('Error: FUNCTION dbinsert tableField is Null');
      var fieldKeyArr = [],
          fieldValueArr = [],
          fieldKey = null,
          fieldValue = null,
          i = 0;

      for (var field in tableField) {
        field.toString();
        tableField[field].toString();
        fieldKeyArr[i] = field;
        fieldValueArr[i] = tableField[field];

        if (typeof fieldValueArr[i] !== 'number') {
          fieldValueArr[i] = '"' + fieldValueArr[i] + '"';
        }

        i++;
      }

      fieldKey = fieldKeyArr.join(',');
      fieldValue = fieldValueArr.join(',');
      var sql = 'INSERT INTO ' + tableName + ' (';
      sql += fieldKey;
      sql += ') ';
      sql += 'VALUES (';
      sql += fieldValue;
      sql += ')';

      _this._dbInfo.transaction(function (ts) {
        console.log('sql', sql);
        ts.executeSql(sql, [], function (ts, result) {
          onSuccess(ts, result);
          console.log('result', result);
        }, function (ts, result) {
          onError(ts, result);
          console.log('result', result);
        });
      });
    });
  }
  /**
   * 
   * 设置存储项目
   * 
   * @param {*} key 键
   * @param {*} value 值
   * @param {*} onSuccess 成功回调
   * @param {*} onError   错误回调
   */


  function setItem$1(key, value, tableName, onSuccess, onError) {
    this._dbInfo.transaction(function (ts) {
      var sql = "INSERT OR REPLACE INTO ".concat(tableName, " ") + '(key, value) VALUES (?, ?)';
      ts.executeSql(sql, [key, value], function (ts, result) {
        onSuccess(ts, result);
      }, function (ts, result) {
        onError(ts, result);
      });
    });
  }
  /**
   * 获取表名
   * @param {*} tableName 表名
   */


  function getItem$1(tableName) {
    var self = this;
    var promise = new Promise(function (resolve, reject) {
      self._dbInfo.transaction(function (ts) {
        try {
          ts.executeSql('select * from ' + tableName, [], function (ts, result) {
            var rst = [];

            if (result) {
              for (var i = 0; i < result.rows.length; i++) {
                rst.push(result.rows.item(i)); // console.info((result.rows.item(i)));
              }
            }

            resolve(rst);
          }, function (ts, message) {
            console.info("查询数据失败！" + message);
          });
        } catch (error) {
          reject(error);
        }
      });
    });
    return promise;
  }
  /**
   * 删除表
   * @param {*} key 值
   * @param {*} tableName 表名
   * @param {*} callback  function 函数
   * @returns 
   */


  function removeItem(key, tableName, callback) {
    var self = this;
    var promise = new Promise(function (resolve, reject) {
      self._dbInfo.transaction(function (ts) {
        try {
          ts.executeSql("DELETE FROM ".concat(tableName, " WHERE id = ?"), [key], function () {
            resolve();
          }, function (t, error) {
            reject(error);
          });
        } catch (error) {
          reject(error);
        }
      });
    });
    return promise;
  }
  /**
   * 更新数据表
   * @param {*} tableName 表明.
   * @param {*} dbParams  sql 拼接参数.
   * @param {*} dbWhere   sql 条件.
   * @returns promise 对象.
   */


  function updateItem(tableName, dbParams, dbWhere) {
    var self = this;
    var SQL = 'UPDATE ' + tableName + ' SET ';
    var paramArr = new Array();
    var paramStr = '';
    var i = 0;

    for (var k in dbParams) {
      if (typeof dbParams[k] !== 'number') {
        dbParams[k] = '"' + dbParams[k] + '"';
      }

      paramArr[i] = k.toString() + '=' + dbParams[k];
      i++;
    }

    paramStr = paramArr.join(',');
    SQL += paramStr;

    if (dbWhere) {
      SQL += ' WHERE ';
      var whereArr = new Array();
      var whereStr = '';
      var n = 0;

      for (var w in dbWhere) {
        if (typeof dbWhere[w] !== 'number') {
          dbWhere[n] = '"' + dbWhere[w] + '"';
        }

        whereArr[n] = w.toString() + '=' + dbWhere[w];
        n++;
      }

      whereStr = whereArr.join(" AND ");
      SQL += whereStr;
    }

    var promise = new Promise(function (resolve, reject) {
      self._dbInfo.transaction(function (ts) {
        try {
          ts.executeSql(SQL, [], function (ts, result) {
            resolve(true);
          }, function (t, error) {
            resolve(false);
          });
        } catch (error) {
          reject(false);
        }
      });
    });
    return promise;
  }
  /**
   * 清空数据表
   * @param {*} tableName 表名
   * @returns 
   */


  function truncateItem(tableName) {
    if (!tableName) {
      console.log('ERROR: Table Name is Null');
      return false;
    }

    var self = this;
    var promise = new Promise(function (resolve, reject) {
      try {
        self._dbInfo.transaction(function (ts) {
          ts.executeSql('delete from ' + tableName, [], function (ts, result) {
            console.log('删除表成功');
            resolve(true);
          });
        }, [], function (ts, result) {
          console.log('DELETE TABLE' + tableName);
          resolve(true);
        }, function (t, error) {
          reject(error);
          console.log(error);
        });
      } catch (error) {
        reject(error);
      }
    });
    return promise;
  }
  /**
   * 删除数据表
   * @param {*} tableName  表名
   * @returns 
   */


  function dropItem(tableName) {
    if (!tableName) {
      console.log('ERROR: Table Name is Null');
      return false;
    }

    var self = this;

    function _drop(tableName) {
      self._dbInfo.transaction(function (ts) {
        ts.executeSql('DROP TABLE' + tableName);
      }, [], function (ts, result) {
        console.log('DROP TABLE' + tableName);
        return true;
      }, function (t, error) {
        console.log('error' + error);
        return false;
      });
    }

    _drop(tableName);
  }

  var webSQLStorage = {
    _driver: 'webSQLStorage',
    _initStorage: _initStorage$1,
    extSetItem: extSetItem,
    getItem: getItem$1,
    setItem: setItem$1,
    removeItem: removeItem,
    updateItem: updateItem,
    truncateItem: truncateItem,
    dropItem: dropItem
  };

  function _initStorage(options) {}
  /**
   * 
   * 设置存储项目
   * 
   * @param {*} key 键
   * @param {*} value 值
   * @param {*} onSuccess 成功回调
   * @param {*} onError   错误回调
   */


  function setItem(key, value, tableName, onSuccess, onError) {
    this._dbInfo.transaction(function (ts) {
      var sql = "INSERT OR REPLACE INTO ".concat(tableName, " ") + '(key, value) VALUES (?, ?)';
      ts.executeSql(sql, [key, value], function (ts, result) {
        onSuccess(ts, result);
      }, function (ts, result) {
        onError(ts, result);
      });
    });
  }
  /**
   * 获取表名
   * @param {*} tableName 表名
   */


  function getItem(tableName) {
    var self = this;
    var promise = new Promise(function (resolve, reject) {
      self._dbInfo.transaction(function (ts) {
        try {
          ts.executeSql('select * from ' + tableName, [], function (ts, result) {
            var rst = [];

            if (result) {
              for (var i = 0; i < result.rows.length; i++) {
                rst.push(result.rows.item(i)); // console.info((result.rows.item(i)));
              }
            }

            resolve(rst);
          }, function (ts, message) {
            console.info("查询数据失败！" + message);
          });
        } catch (error) {
          reject(error);
        }
      });
    });
    return promise;
  }

  var webIndexDbStorage = {
    _driver: 'webIndexDbStorage',
    _initStorage: _initStorage,
    getItem: getItem,
    setItem: setItem
  };

  var isArray = Array.isArray || function (arg) {
    return Object.prototype.toString.call(arg) === '[object Array]';
  };

  var DefinedDrivers = {};
  var DefaultDrivers = {
    INDEXDDB: webIndexDbStorage,
    WEBSQL: webSQLStorage
  };
  var DefaultDriverOrder = [DefaultDrivers.INDEXDDB._driver, DefaultDrivers.WEBSQL._driver];
  var DefaultConfig = {
    description: 'localStore',
    driver: DefaultDriverOrder.slice(),
    name: 'localStore',
    size: 4980736,
    storeName: 'master',
    version: 1.0
  };
  var DriverSupport = ['webSQLStorage', 'asyncStorage'];

  function extend() {
    for (var i = 1; i < arguments.length; i++) {
      var arg = arguments[i];

      if (arg) {
        for (var key in arg) {
          if (arg.hasOwnProperty(key)) {
            if (isArray(arg[key])) {
              arguments[0][key] = arg[key].slice();
            } else {
              arguments[0][key] = arg[key];
            }
          }
        }
      }
    }

    return arguments[0];
  }

  var LocalStore = /*#__PURE__*/function () {
    function LocalStore(options) {
      _classCallCheck(this, LocalStore);

      for (var driverTypeKey in DefaultDrivers) {
        if (DefaultDrivers.hasOwnProperty(driverTypeKey)) {
          var driver = DefaultDrivers[driverTypeKey];
          var driverName = driver._driver;
          this[driverTypeKey] = driverName;
        }
      }

      this._defaultConfig = extend({}, DefaultConfig);
      this._config = extend({}, this._defaultConfig, options);
      this._driverSet = null;
      this._initDriver = null;
      this._ready = false;
      this._dbInfo = null;
    }

    _createClass(LocalStore, [{
      key: "defineDriver",
      value: function defineDriver(driverObject, callback, errorCallback) {
        this._extend(driverObject);
      }
    }, {
      key: "_extend",
      value: function _extend(libraryMethodsAndProperties) {
        extend(this, libraryMethodsAndProperties);
      }
    }, {
      key: "config",
      value: function config(options) {
        this._config = options;
      }
    }, {
      key: "supports",
      value: function supports(driverName) {
        return DriverSupport.filter(function (item) {
          return item === driverName;
        }) ? true : false;
      }
    }, {
      key: "_getSupportedDrivers",
      value: function _getSupportedDrivers(drivers) {
        var supportedDrivers = [];

        for (var i = 0, len = drivers.length; i < len; i++) {
          var driverName = drivers[i];

          if (this.supports(driverName)) {
            supportedDrivers.push(driverName);
          }
        }

        return supportedDrivers;
      }
    }, {
      key: "setDriver",
      value: function setDriver(drivers, callback, errorCallback) {
        var self = this;

        if (!isArray(drivers)) {
          drivers = [drivers];
        } // 验证是否支持驱动类型.


        var supportedDrivers = this._getSupportedDrivers(drivers);

        function initDriver() {
          if (supportedDrivers.indexOf('webSQLStorage') > -1) {
            // 核心的方法就是实例合并给 extend
            extendSelfWithDriver(webSQLStorage);
          }

          if (supportedDrivers.indexOf('asyncStorage') > -1) {
            extendSelfWithDriver(webIndexDbStorage);
          }
        }

        function extendSelfWithDriver(driver) {
          self._extend(driver);

          self._initStorage(self._config);
        }

        initDriver();
      }
    }, {
      key: "driver",
      value: function driver() {
        return this._driver || null;
      }
    }, {
      key: "getDriver",
      value: function getDriver(driverName) {
        return DefinedDrivers[driverName];
      }
    }, {
      key: "ready",
      value: function ready(callback) {
        var promise = new Promise(function (resolve, reject) {
          resolve();
        });
        return promise;
      }
    }]);

    return LocalStore;
  }();

  var localstore = new LocalStore();

  return localstore;

})));
